### ISSUE TITLE
research(v2.4.0): implement a bounded recurrent marked temporal point-process proposal engine

### ISSUE THREAD

**Issue by dmidlo** [dmidlo/histdata.com-tools#453]
## Outcome

Add and evaluate one bounded CPU-only recurrent marked temporal point-process challenger after the empirical product path and the transparent #450–#452 statistical challengers have established the shared contracts and controls.

Parent: #430  
Dependencies satisfied: #431, #432, #433, #436, #449, #450, #451, #452

## Current-code integration decision

The neural challenger is a separate, opt-in family. It must not alter the fixed `EventClockFamily`, `HawkesExcitationStructure`, or `RegimeHawkesModulation` registries, introduce a heavyweight neural runtime into core dependencies, change the certified empirical default, or bypass the generator-neutral carving/certification path.

The implementation gets its own versioned dataset, configuration, training/checkpoint, fit, generation/evidence, lineage, candidate, and benchmark-adapter contracts while reusing:

- calibration-only `EventClockCalibrationWindowV1` event inputs;
- synchronized `BenchmarkEventV1` generation and reverse-degradation scoring;
- `ReconstructionRunV1` / `ReconstructionWindowV1` ownership;
- exact immutable-anchor and full-input content validation;
- generator-neutral `ReconstructionCandidateBatchV1` carving;
- v2 technological feed-epoch transition/uncertainty context; and
- the existing v2.3.0 certification boundary.

## Fixed architecture and rationale

Implement exactly one small recurrent marked temporal point process, `rmtpp_cpu_v1`, based on Du et al., *Recurrent Marked Temporal Point Processes: Embedding Event History to Vector* (KDD 2016):

https://doi.org/10.1145/2939672.2939875

For event-history state `h_i`, positive elapsed time `tau`, and next joint mark `k`:

```text
h_i = tanh(W_h h_(i-1) + W_x x_i + b_h)
lambda(t_i + tau | H_i) = exp(b_lambda + v^T h_i + w * tau)
p(mark = k | H_i) = softmax(W_mark h_i + b_mark)[k]
```

The joint mark is destination symbol × quote transition (`ask_only`, `bid_only`, `joint`, `unchanged`). The time likelihood uses the exact closed-form compensator of the declared positive-`w` RMTPP intensity, and sampling uses its exact inverse CDF. No quadrature, rejection sampler, or hidden continuous-time approximation is permitted.

This architecture is selected over a continuous-time LSTM Neural Hawkes model because its likelihood and inverse sampler are exact, its state transition is compact enough for a dependency-free deterministic CPU implementation, and every parameter/gradient/resource seam can be reviewed. Neural Hawkes remains methodological context, not a second implementation:

https://arxiv.org/abs/1612.09328

The mark head is conditionally independent of the next elapsed time given `h_i`; that restriction is explicit and is not described as state of the art.

## Dataset, split, context, and leakage contract

- [x] A content-addressed dataset manifest records every source window, split role, event-content digest, context digest, sequence length, symbol/mark support, and ordering policy without embedding protected holdout rows.
- [x] Dataset construction is point-in-time safe, resets recurrent state at every window boundary, and rejects overlapping/duplicated window identities or events outside their declared window.
- [x] Training and tuning are split at whole-window boundaries before preprocessing; untouched benchmark validation/final-holdout rows never enter training, tuning, normalization, early stopping, or checkpoint selection.
- [x] The real corpus uses the first calibration occurrence of each session for training and the second for tuning, preserving all three sessions in both roles.
- [x] Exact sequence hashes and a fixed near-duplicate signature audit run across train/tune and protected benchmark splits; overlap, exact duplicates, or a predeclared near-duplicate collision fails closed.
- [x] Latent neural state, technological feed epoch/transition, session, benchmark split, and optional observed context remain distinct versioned axes.
- [x] Stable epoch versus transition assignment, boundary identity/support/uncertainty, context content, availability time, and use time are preserved exactly.
- [x] Ex-ante construction requires a valid as-of boundary; ex-post construction forbids one. Future context or events cannot enter an ex-ante model.

## Preprocessing and mark semantics

- [x] Events are ordered deterministically by time, symbol, source sequence, and event identity; equal-time ordering is retained rather than collapsed.
- [x] Positive inter-event durations, window-start duration, log-duration transform, training-only normalization, start token, symbol vocabulary, quote-transition vocabulary, and prior-quote reset policy are versioned.
- [x] Input/event/mark vocabularies are fixed; unsupported symbols, nonpositive time reversal, nonfinite quotes, invalid spreads, or unknown marks fail closed.
- [x] Preprocessing statistics are learned from training only, serialized in the checkpoint, bounded, finite, and identity-checked at fit and adapter binding.
- [x] Dataset/preprocessing round trips are strict, content-addressed, and deterministic.

## Training, checkpoint, and fit contract

- [x] The implementation is standard-library, CPU-only, full-batch deterministic backpropagation through time with a fixed hidden width, initialization seed, learning rate, gradient clip, parameter clip, epoch limit, and early-stopping rule.
- [x] Accelerator count is fixed to zero; a GPU/accelerator request refuses before dataset materialization or training.
- [x] Hardware/runtime evidence records OS, architecture, Python implementation/version, CPU-only policy, seed, and deterministic-math limitations without making volatile metadata part of scientific identity.
- [x] The exact normalized train/tune time and mark negative log likelihood, tune time error, mark accuracy, gradient norm, selected epoch, and complete bounded loss trace are recorded.
- [x] Checkpoints include architecture, vocabularies, preprocessing statistics, all finite parameter tensors, optimizer-free replay state, epoch, metrics, and an immutable content digest.
- [x] Checkpoint selection uses tuning loss only; no validation/final-holdout metric can select or mutate a checkpoint.
- [x] Fit/config/dataset/training/checkpoint identities are revalidated at construction, serialization, adapter binding, generation, and carving.
- [x] Nonfinite loss/gradient/parameter, exploding gradients beyond clipping policy, empty support, missing marks/symbols, no improving checkpoint, or checkpoint tamper exposes no usable model.
- [x] Event, window, sequence, feature, mark, parameter, diagnostic, epoch, gradient-work, checkpoint-byte, memory, and wall-time envelopes are independent and bounded.
- [x] Failed/refused training is deterministic and exposes no partial checkpoint or parameters.

## Generation and lineage contract

- [x] Generation uses one synchronized recurrent state, deterministic semantic seed, exact RMTPP inverse-CDF times, and softmax joint-mark sampling.
- [x] Explicit prior history is synchronized, strict-prior, cardinality/age bounded, and outside-lookback history contributes nothing.
- [x] Observed anchors are consumed only when generation reaches their time; future anchors cannot update hidden state early.
- [x] Generated candidates update recurrent state only after their event, while observed anchors remain byte-for-byte unchanged.
- [x] Candidate times are unique and strictly between observed destination anchors; a mark targeting an unsupported interval is skipped/accounted or causes a declared whole-attempt refusal, never silently clamped.
- [x] Quote projection, mark probability, elapsed-time density/intensity, hidden-state digest, source/parent, anchor interval, and lineage are deterministic and auditable.
- [x] Sample-step, generated-event, per-interval, amplification, history, parameter, memory, output-byte, and runtime limits refuse the entire attempt without truncation or partial output.
- [x] Successful/empty attempts bind full input-event, anchor, retained-history, context, config, fit, dataset, training, and checkpoint digests.
- [x] Failed/refused attempts expose no candidate rows or lineage and are non-carveable.

## Carving and benchmark seams

- [x] Neural candidate batches satisfy `ReconstructionCandidateBatchV1` and revalidate run/window/member/config/fit/checkpoint/generation/context/anchor identities.
- [x] The shared historical carving engine remains authoritative for anchor integrity, resources, context support, quarantine, conditioned intensity, spread projection, synchronized cross-series validation, broker conditioning, and final quality.
- [x] The reverse-degradation campaign accepts either no neural config or exactly the fixed `rmtpp_cpu_v1` config.
- [x] The neural challenger runs beside the empirical motif, four #450 clocks, three #451 static Hawkes models, and two #452 regime Hawkes models over identical corpus/splits/scenarios/members/anchors.
- [x] Candidate metrics expose dataset/split/leakage, train/tune likelihood and calibration, checkpoint/determinism/resource evidence, plus timing, burst/quiet, dispersion, duration, marks, spread/tails, refusal, and anchor integrity.
- [x] Incremental evidence is directly inspectable against empirical motif, static full Hawkes, and regime baseline+excitation.
- [x] Claims require all three symbols, all sessions, multiple technological epochs/transition strata, untouched holdouts, and more than likelihood improvement.
- [x] Campaign artifacts are bounded, row-free, content-addressed, replay-verified, and declare `automatic_winner: false`.
- [x] This issue cannot promote the neural challenger or alter the production default.

## Verification

- [x] Controlled fixtures recover learnable sequence/mark dependence and show lower held-out NLL than an untrained deterministic checkpoint.
- [x] Tests cover strict round trips, split/leakage isolation, preprocessing, equal-time ordering, BPTT gradients, finite-difference spot checks, deterministic training/checkpoint selection, tamper, resources, history, anchors, generation, lineage, carving, and no-config coexistence.
- [x] Focused integration proves coexistence with empirical, clock, static-Hawkes, and regime-Hawkes challengers.
- [x] A real-corpus closure campaign reports all eleven challenger variants (empirical + four clocks + three static Hawkes + two regime Hawkes + one neural TPP) with no automatic winner.
- [x] Real artifacts replay with exact hashes and report train/tune/validation/final-holdout, session, event-state, symbol, and technological-epoch/transition coverage.
- [x] Full test suite, warnings-as-errors docs build, and all-files pre-commit hooks pass before commit/push/closure.
- [x] Closure evidence records the exact commit, test counts, artifact identities/digests, resource measurements, train/tune/checkpoint results, holdout/refusal results, cross-epoch evidence, and campaign decision.

## Scientific nonclaims and non-goals

- The recurrent hidden state is not an observed economic regime or causal explanation.
- Neural complexity or parameter count is not evidence of improvement.
- This is not a continuous-time LSTM Neural Hawkes implementation.
- The conditionally independent mark head is not claimed to be universally adequate.
- This does not implement broker-specific adaptation, diffusion generation, or Add-and-Thin.
- This does not replace the certified v2.3.0 default.
- This does not create a standalone package release solely for this issue.

## SemVer

Additive optional minor feature for the v2.4.0 reconstruction-generator challenger train.

**Comment by dmidlo** [comment]
## Closure evidence

Implemented, validated, and pushed in commit 68474fd806ae51bb2e003b23693b13caebec2d0c on dev. Local HEAD and origin/dev are exact, and the worktree is clean.

### Verification

- Acceptance audit: 48/48 criteria verified; the issue checklist is checked.
- Full pytest: 2,129 passed, 4 expected environment-dependent skips, 0 failed, 0 errors in 1,623.39 seconds. Retained JUnit reports 2,133 collected tests and 1,622.936 seconds.
- Focused neural/campaign tests: strict round trips and identities; train-only preprocessing; controlled held-out NLL and mark-NLL improvement; equal-time ordering; exact censoring; analytic BPTT gradient against finite difference; protected duplicate and ex-ante leakage refusal; deterministic generation, anchors, history and resource closure; tamper/context mismatch; shared carving; no-config/fixed-config campaign coexistence.
- Documentation: Sphinx 8.1.3 warnings-as-errors build passed for all 39 sources.
- Repository gates: all-files pre-commit passed, including Black, Ruff, mypy over 156 source files, Pyroma, shellcheck, actionlint, package smoke, and architecture diagrams. Commit and pre-push hooks passed again.

### Retained real-corpus evidence

Campaign: reverse-degradation-campaign:sha256:eb9741d3b1d5bb4678eca894685a216870e67f1fb103d18e0520dfb58fab820e  
Corpus: reverse-degradation-corpus:sha256:a760a010d44de2d6258b7c3d71651b00bc24eaef53092f37bd75b3ae2395c5dc  
Motif index: reference-motif-index:sha256:b5d5e7d9580fac375c42677fe5d03be96fafc190f799364a52566af7aa5a2589

The campaign replayed all nine source inputs and 54 symbol-window partitions, produced all 15 reports and all 11 challengers, covered 18 measured windows with two ensemble members, and retained automatic_winner=false. Campaign integrity passed with zero source-hash, information-audit, holdout-neighbor, dense-identity, degradation, or anchor violations.

Neural identities:

- config: neural-tpp-config:sha256:64959295a3c196a5ef0f150f6b5f8fc7fba7c79081c96a17b7336dad2b491e81
- dataset: neural-tpp-dataset:sha256:19951c2f97711ffde5f2951a19696b1c9130d1cffa666484d40314c65c953b0b
- training: neural-tpp-training:sha256:c99fc606e4985e57fc2b5c8aa052140960ba50b6d9d43e0170585d8d57876e6f
- checkpoint: neural-tpp-checkpoint:sha256:02ea68506721c5e122ad62e7eba2ed40f4282dccd93282f2f946d53aa78bd3fc
- fit: neural-tpp-fit:sha256:801842ad365e82349ea04cb14b53b2ca3d0a498f89eeaa1bff4076acc585aa17
- candidate: benchmark-candidate:sha256:90b2a9242dca2b2750eb4953b923090388c700ea4b4fb536ce9e16d1c6813332
- report: reverse-degradation-candidate-report:sha256:c4120afc4d3748efc0e680a2c71b667eabc338338b0dd26bf574dc0bfb272349

The fit used three training windows / 1,835 events and three tuning windows / 1,687 events. Epoch 40 of 40 was selected. Train NLL was 3.461239 (time 0.982061, mark 2.479178); tune NLL was 3.549778 (time 1.070875, mark 2.478903), tune mark accuracy 0.221695, log-duration RMSE 0.592358, and mean PIT 0.375962. The checkpoint contains 309 parameters / 6,750 bytes. Gradient work was 22,680,600, maximum gradient norm 0.279580, estimated fit peak memory 4,221,696 bytes, and exact/near-duplicate/overlap counts were all zero.

The neural candidate evaluated all 12 validation/final windows with two members: 0 failures, 0 refusals, and 0 immutable-anchor violations. Its mean stream errors were event count 0.381555, interarrival histogram L1 0.210748, update-transition L1 0.591464, path variation 0.303450, spread tail 0.025364, and count dispersion 0.587162.

The corpus covered all three symbols and sessions and the declared epoch-03 calibration to epoch-04 evaluation boundary. No feed-transition window was selected by this real corpus; the transition contract is covered by focused fixtures, and no cross-transition improvement is claimed.

### Resource and artifact closure

Campaign runtime was 38.358593 seconds, process peak RSS 159,088,640 bytes, and retained artifact size 954,190 bytes.

Retained file SHA-256 digests:

- scorecard: fcd26e11535fe9fd499d0fa88f0bcf388fe7c0367a117a770056790debc88ada
- manifest: d1ddf45d68ade8c1ba4abc3df5a60a26483bb3eab950d4c29f53709e9214ed24
- motif artifact: 048dd46deabf66643fc55b9cb2a996828c88f8c099a9d9573a4d29a632bce9a3
- leakage audit: e1cd0fcc9da45756dfe5f121742bfd26e1ba0cdda46971fbae7eb691402d0165
- resource audit: 173eb81b97f612a5937e3d8b29ddbbaf99c4ea584924a30e84e2362325c0f265

### Decision

The RMTPP is scientifically inspectable and operationally bounded, but it did not clear the candidate promotion gates for event count, path variation, or update-transition fidelity. It remains an opt-in research challenger. No production default, certified empirical path, fixed clock/Hawkes registry, broker adaptation, or release version changed. Closing #453 as implemented and evaluated, with no model promotion.

**Comment by dmidlo** [comment]
## 2026-07-30 classification correction

This completed module is a proposal-engine implementation for #430. “Challenger” was only its role in the retained benchmark campaign. It is not an optional post-objective feature, and implementation does not imply promotion. #489 owns first-party reachability; #490 owns powered qualification. Its current non-promotion result is retained scientific evidence.
